### PR TITLE
Fix ConcurrentModificationException due to FrameMetricsAggregator manipulation

### PR DESCRIPTION
**Pull request by adinauer**
androidx.core.app.FrameMetricsAggregator is not thread safe, calling .remove() from a background thread may cause a
ConcurrentModificationException

## :scroll: Description
Ensure calling .remove() is done on the UI thread.

## :bulb: Motivation and Context
SentryTracer uses a `Timer` and `TimerTask` to finish open Transactions on a background thread. Finishing a transaction also triggers the calculation of the frame metrics as well as unregistering the activity from metric collection (`FrameMetricsAggregator.remove(activity)`), FrameMetricsAggregator is not thread safe, causing sporadic `ConcurrentModificationException`s.

## :green_heart: How did you test it?
Ensured current tests are not breaking, otherwise it's hard to test a ConcurrentModificationException.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:35]: Get it via ctor to make it testable.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:73]: Let's add a comment and log it out with the `logger`.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:229]: `reset` can be out of both blocks, it does not touch `OnFrameMetricsAvailableListener` and it copies the instance of `mMetrics`

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:65]: `AppLifecycleIntegrationTest` has tests that it checks if its executed in the main thread or not.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java
Comment (markushi) [reply sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java:102]: I agree, I still added it in order to make all `frameMetricsAggregator.x` calls safer. Yes, it's safe now but in theory it could change with a future update to the androidx lib. But I'm fine with removing it 👍

File: sentry-android-core/api/sentry-android-core.api
Comment (markushi) [inline sentry-android-core/api/sentry-android-core.api:4]: @marandaneto I added the handler argument to both the test and the non-test ctor, this now changes the .api file - what do you think, should the handler be an argument for the `@TestOnly` constructor or is it fine this way?

Comment (marandaneto) [reply sentry-android-core/api/sentry-android-core.api:4]: It's fine IMO.

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt
Comment (marandaneto) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:355]: You can omit `times(1)` because its the default.

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt
Comment (marandaneto) [inline sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt:363]: I'd prefer one test per use case.

File: sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java]: Your call, upgrades to the androidx lib can cause new issues, I agree.

File: CHANGELOG.md
Comment (adinauer) [inline CHANGELOG.md:3]:
```suggestion
## Unreleased

### Fixes
```

**Review by adinauer — COMMENTED** [review]
Feel free to merge. If we can extract some of the duplicated code to make it easier to read / maintain I'd vote for that.

File: sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt
Comment (philipphofmann) [reply sentry-android-core/src/test/java/io/sentry/android/core/ActivityFramesTrackerTest.kt]: ❤️ , you still remember after 2,5 years @marandaneto 😄

**Review by philipphofmann — APPROVED** [review]
LGTM

**Comment by markushi** [comment]
I've updated the description (copied it over from the initial PR)

**Comment by marandaneto** [comment]
> Seeing this finally getting ready for merge, the changes are functional but definitely not very readable/maintainable. I was wondering if we should introduce a helper function do deal with the `try/catch` and ensure being run on the main thread.
> 
> E.g.
> 
> ```
> 
> void runSafelyOnMainThread(Runnable runnable, String tag) {
>     try {
>         if (MainThreadChecker.isMainThread()) {
>             runnable.run();
>         } else {
>             handler.post(
>                 () -> {
>                 try {
>                     runnable.run();
>                 } catch (Throwable t) {
>                     if (logger != null) {
>                     logger.log(SentryLevel.ERROR, "Failed to run " + tag, t);
>                     }
>                 }
>                 });
>         }
>     } catch (Throwable t) {
>         if (logger != null) {
>             logger.log(SentryLevel.ERROR, "Failed to run" + tag, t);
>         }
>     }
> }
> 
> ...
> runSafelyOnMainThread(() -> { frameMetricsAggregator.stop(); }, "frameMetricsAggregator.stop");
> ```

Makes sense, and I'd reuse where we already do something similar.

**Comment by adinauer** [comment]
> Makes sense, and I'd reuse where we already do something similar.

Can we do that as a separate issue / PR as not to block this from being released?

**Review by markushi — APPROVED** [review]

**Pull request merged**
